proposals/utils/pdf_diff_logic.py

```diff
@@ -0,0 +1,1132 @@
+
+from datetime import date
+from copy import copy
+
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.template.loader import get_template
+from django.utils.html import format_html
+from django.utils.safestring import mark_safe
+from django.utils.translation import ugettext as _
+
+from proposals.templatetags.proposal_filters import needs_details, medical_traits, \
+necessity_required, has_adults
+from proposals.templatetags.diff_tags import zip_equalize_lists
+
+'''NOTE: The current implementation does not account for page break formatting, present 
+in the previous pdf template
+NOTE: As of now, the diff has no workaround for when observations or interventions are of a
+different version, in the different proposals.'''
+
+class PDFSection:
+    '''This is the main class, from which most classes are constructed.
+    It can take one object or two objects in a tuple, and produce html tables accordingly.
+    The most important methods are render and make_rows. Make rows makes use of get_row_field,
+    which is mostly overwritten for specific sections.
+    If this class is used in the diff, object should be the parent object and object_2 the 
+    revision/amendment.'''
```

Comment:
Tacking on to this, as I have an alternative solution that might be interesting.

What we're seeing here in `PDFSection` and also in the `ObservationSection` is _type keying_, which tends to happen when a class or function fulfills two different functions resulting in a kind of split personality. In this case the class keeps checking if it has an extra object to determine what template it needs, what rows it desires, what it wants for dinner, etc. It is usually best to split such classes up, as Ty says.

In my opinion you have a perfectly good `PDFSection` for the PDF case. And simply put, a diff is just two of those. Why not have a standalone `DiffSectionV2` class, that takes two PDF sections as arguments. Specifically I mean that `DiffSectionV2` takes two instantiated section objects[1] and not two proposal or study (or whatever) objects. By standalone I mean that it would not inherit from `PDFSection`, or anything necessarily.

I think this design would be more resilient than passing two objects to a `PDFSection` spinoff class and then making it do everything twice. Let the two sections do their own thing as if they're just making a PDF and then combine them later. You already have most of this code hidden under diff-checking if statements in `PDFSection` waiting to be cut and pasted.

I must say that I find the multiple inheritance solution (e.g. `StudySectionDiff`) quite creative and cool. But sadly that's how you get MRO problems.

[1] Or maybe None to indicate a missing section. There are probably multiple ways to sort out the missing section problem here, do what feels right.